pkg/binary/binary.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package binary
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//nolint
+const (
+	// GOOS labels
+	LINUX  = "linux"
+	DARWIN = "darwin"
+	WIN    = "windows"
+
+	// GOARCH Architecture labels
+	I386    = "386"
+	AMD64   = "amd64"
+	ARM     = "arm"
+	ARM64   = "arm64"
+	PPC     = "ppc"
+	PPC64LE = "ppc64le"
+	S390    = "s390"
+	RISCV   = "riscv"
+)
+
+// Binary is the base type of the package. It abstracts a binary executable
+type Binary struct {
+	options *Options
+	binaryImplementation
+}
+
+// Options to control the binary checker
+type Options struct {
+}
+
+// DefaultOptions set of options
+var DefaultOptions = &Options{}
+
+// New creates a new binary instance.
+func New(filePath string) (bin *Binary, err error) {
+	// Get the right implementation for the specified file
+	return NewWithOptions(filePath, DefaultOptions)
+}
+
+// NewWithOptions creates a new binary with the specified options
+func NewWithOptions(filePath string, opts *Options) (bin *Binary, err error) {
+	// Get the right implementation for the specified file
+	impl, err := getArchImplementation(filePath, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting arch implementation")
+	}
+	return &Binary{
+		options:              opts,
+		binaryImplementation: impl.(binaryImplementation),
```

Comment:
(non blocking) I would prefer having a typed `binaryImplementation` here with avoiding of casting the `interface{}`.

Reply:
Done, me too :)